### ISSUE TITLE
'FakeItEasy.Analyzer.RepeatedAssertionAnalyzer' threw an exception of type 'System.NullReferenceException'

### ISSUE THREAD

**Issue by Mertsch** [FakeItEasy/FakeItEasy#1328]
FIE v4.5.0
The `CS8032` disappeared after restarting VS 2017 already upgraded to 4.5.0 (from 4.4.0)

```
Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Warning	AD0001	Analyzer 'FakeItEasy.Analyzer.RepeatedAssertionAnalyzer' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.	XXX.Tests	D:\Source\XXX.Tests\CSC	1	Active
Warning	CS8032	An instance of analyzer FakeItEasy.Analyzer.RepeatedAssertionAnalyzer cannot be created from C:\Users\XXX\.nuget\packages\fakeiteasy.analyzer.csharp\4.5.0\analyzers\dotnet\cs\FakeItEasy.Analyzer.CSharp.dll: Could not load type 'FakeItEasy.Analyzer.RepeatedAssertionAnalyzer' from assembly 'FakeItEasy.Analyzer.CSharp, Version=4.0.0.0, Culture=neutral, PublicKeyToken=eff28e2146d5fd2c'..	XXX.Tests		1	Active	System.TypeLoadException: Could not load type 'FakeItEasy.Analyzer.RepeatedAssertionAnalyzer' from assembly 'FakeItEasy.Analyzer.CSharp, Version=4.0.0.0, Culture=neutral, PublicKeyToken=eff28e2146d5fd2c'.
   at System.Reflection.RuntimeAssembly.GetType(RuntimeAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type)
   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.Extensions`1.GetAnalyzersForTypeNames(Assembly analyzerAssembly, IEnumerable`1 analyzerTypeNames, Boolean& reportedError)
-----


```

**Comment by thomaslevesque** [comment]
Thanks @Mertsch for reporting this.
Could you show the code that makes the analyzer throw? It would be helpful to reproduce the problem.

**Comment by Mertsch** [comment]
@thomaslevesque
I have no idea how to know which test caused this exception (apparently not all) and I can't share the source

**Comment by Mertsch** [comment]
@thomaslevesque I have tried step by step and nailed down the issue to
```
(await Assert.ThrowsExceptionAsync<InvalidOperationException>(() => { })))
                .Message.ShouldContain(nameof(ILaundryAnalyzer.InitializeAsync));
//                                                             ^^^^^^^^^^^^^^^
```
```
[UsedImplicitly]
public interface ILaundryAnalyzer
{
    int StepId { get; }

    [NotNull]
    Task<bool> InitializeAsync(int projectId, CancellationToken ct);
}
```

`nameof(ILaundryAnalyzer.StepId)` does not throw the exception

**Comment by thomaslevesque** [comment]
Thanks @Mertsch, that's helpful.
I'm not sure what's going on here, but we'll look into it.

**Comment by thomaslevesque** [comment]
OK, I got it. In this method:

```
private static string GetTypeName(SyntaxNodeAnalysisContext context, ArgumentSyntax argument)
{
    return context.SemanticModel.GetTypeInfo(argument.GetExpression()).ConvertedType.ToString();
}
```
`ConvertedType` can be null, if the expression has no type. For instance, in `nameof(ILaundryAnalyzer.InitializeAsync)`, the expression `ILaundryAnalyzer.InitializeAsync` has no type.

The weird thing is that we shouldn't reach the point where we call `GetTypeName`: the analyzer is called only on invocation expressions, and `ILaundryAnalyzer.InitializeAsync` shouldn't be one, since it's not invoked here. And anyway, we even check that the argument list has 1 item, and there's none.

**Comment by thomaslevesque** [comment]
OK, I understand now. It's not `ILaundryAnalyzer.InitializeAsync` that is considered an invocation expression, but `nameof(ILaundryAnalyzer.InitializeAsync)`. `nameof` is a contextual keyword; it's considered a keyword only if there's no method called "nameof". From a syntactic point of view, `nameof(ILaundryAnalyzer.InitializeAsync)` *is* a invocation expression: it will resolve to a method call if a method called "nameof" is in scope.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.5.1](https://github.com/FakeItEasy/FakeItEasy/releases/4.5.1).
Thanks, @Mertsch. Look for your name in the release notes! :trophy:
